lib/SILGen/SILGenApply.cpp

```diff
@@ -569,6 +569,20 @@ class Callee {
     if (Constant)
       constant = Constant;
 
+    auto isDistThunkCall = false;
+    if (constant) {
+      auto func = constant->getFuncDecl();
+      if (func && func->isDistributed()) {
+        fprintf(stderr, "\n[%s:%d] (%s) decl is dist: %d\n", __FILE__, __LINE__, __FUNCTION__, func->isDistributed());
+        constant->dump();
+
+        if (auto proto = dyn_cast<ProtocolDecl>(func->getDeclContext())) {
+          isDistThunkCall = func->isDistributed() &&
+                            isa<ProtocolDecl>(func->getDeclContext());
```

Comment:
This whole `&&` is redundant, because we've already checked both `func->isDistributed()` (outer `if`) and that it's in a protocol (inner `if`). But if you take my ABI suggestion above, we'll check for "not `async throws`" here.

Reply:
Ok right the isa is unnecessary, the other part... I'm not sure... I think that'd eventually work if we go all the way with some more refactoring 🤔 Not quite there yet.